src/engine/strat_engine/metadata.rs

```diff
@@ -4,262 +4,365 @@
+                    try!(f.read_exact(&mut json_buf));
+
+                    if json_crc != crc32::checksum_ieee(&json_buf) {
+                        return Err(EngineError::Engine(ErrorEnum::Invalid,
+                                                       "MDA region json CRC".into()));
+                    }
+                    Some(json_buf)
+                }
+            };
+
+            Ok((mda, json_buf))
+        };
+
+        let (mda0, json0) = load_region(0)
+            .unwrap_or_else(|_| load_region(2).unwrap_or_else(|_| (MDAHeader::new(), None)));
+        let (mda1, json1) = load_region(1)
+            .unwrap_or_else(|_| load_region(3).unwrap_or_else(|_| (MDAHeader::new(), None)));
+
+        let latest_json = match mda0.last_updated.cmp(&mda1.last_updated) {
+            Ordering::Less => json0,
+            Ordering::Greater => json1,
+            Ordering::Equal => json0,
+        };
+
+        Ok(MDARegions {
+            region_size: region_size,
+            mdas: [mda0, mda1],
+            latest_json: latest_json,
+        })
+    }
+
+    pub fn save(&mut self, time: &Timespec, data: &[u8], f: &mut File) -> EngineResult<()> {
+        let region_size: u64 = *self.region_size * SECTOR_SIZE;
+
+        let mut save_region = |region: usize| -> EngineResult<()> {
+            let hdr_buf = MDAHeader::to_buf(data, time);
+            let offset = BDA_STATIC_HDR_SIZE + (region as u64 * region_size);
+
+            if MDA_REGION_HDR_SIZE + data.len() > region_size as usize {
+                return Err(EngineError::Engine(ErrorEnum::Invalid,
+                                               "data larger than region_size".into()));
             }
+
+            try!(f.seek(SeekFrom::Start(offset)));
+            try!(f.write_all(&hdr_buf));
+            try!(f.write_all(data));
+            try!(f.flush());
+
+            Ok(())
         };
-        Ok(ownership)
+
+        let older_region = self.older();
+
+        // Save to primary and backup regions
+        // TODO: Should we ignore errors?
+        try!(save_region(older_region));
+        try!(save_region(older_region + 2));
+
+        self.mdas[older_region].last_updated = *time;
+        self.mdas[older_region].used = data.len() as u64;
+        self.latest_json = Some(data.to_vec());
+
+        Ok(())
+    }
+
+    pub fn older(&mut self) -> usize {
+        match self.mdas[0].last_updated.cmp(&self.mdas[1].last_updated) {
+            Ordering::Less => 1,
+            Ordering::Greater => 0,
+            Ordering::Equal => 1,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct MDAHeader {
+    pub last_updated: Timespec,
+    pub used: u64,
+}
+
+impl MDAHeader {
+    pub fn new() -> MDAHeader {
+        MDAHeader {
+            last_updated: Timespec::new(0, 0),
+            used: 0,
+        }
+    }
+
+    pub fn from_buf(buf: &[u8; MDA_REGION_HDR_SIZE]) -> EngineResult<(MDAHeader, u32)> {
```

Comment:
This could definitely use a header, if only to clarify what the second value returned in the result is.